tensorflow/core/framework/device_factory.cc

```diff
@@ -63,17 +64,32 @@ int32 DeviceFactory::DevicePriority(const string& device_type) {
 }
 
 // static
-void DeviceFactory::Register(const string& device_type, DeviceFactory* factory,
-                             int priority) {
+// If subdevice type is not found in the device factories, that means
+// device is not registered. In this case, device type will be used as
+// subdevice type.
+string DeviceFactory::SubDeviceType(const string& device_type) {
+  tf_shared_lock l(*get_device_factory_lock());
+  std::unordered_map<string, FactoryItem>& factories = device_factories();
```

Comment:
done, I change it to const &, thanks

Reply:
Sorry I meant `factories` should be a const ref, not the return value.  I don't think returning a const ref string is safe since e.g. resizing the hash table will invalidate the string.